server/dash/dashboards/ncSumComp.py

```diff
@@ -0,0 +1,407 @@
+    df.loc[:, "timeToClose"] = df.loc[:, "timeToClose"].fillna(0.0000001)
+
+    # Replace negative values
+    # TODO: figure out what to do when there is no data avaialble
+    df = df[df["timeToClose"] > 0]
+    if df.shape[0] == 0:
+        raise PreventUpdate()
+    else:
+        q3, q1 = np.percentile(df.loc[:, "timeToClose"].astype(int), [75, 25])
+        iqr = q3 - q1
+        if not iqr:
+            numBins = 100
+        else:
+            numBins = int((2 * iqr) / (df.shape[0]**(1 / 3)))
+
+        # Log Transform, Compute IQR, then exclude outliers
+        df.loc[:, "logTimeToClose"] = np.log(df.loc[:, "timeToClose"])
+        log_q3, log_q1 = np.percentile(df.loc[:, "logTimeToClose"], [75, 25])
+        log_iqr = log_q3 - log_q1
+
+    # Data Quality switch to remove outliers as defined by Median +- 1.5*IQR
+    if data_quality_switch:
+        # TODO: figure out what happens when the filtering mechanism output no data at all
+        temp = df[(df.loc[:, "logTimeToClose"] > 1.5 * log_iqr - np.median(df.loc[:, "logTimeToClose"])) &
+                   (df.loc[:, "logTimeToClose"] < 1.5 * log_iqr + np.median(df.loc[:, "logTimeToClose"]))]
+        if temp.shape[0] > 0:
+            df = temp
+        data_quality_output = "Quality Filter: On"
+    else:
+        data_quality_output = "Quality Filter: Off"
+
+    # Distribution for the total number of requests
+    print(" * Generating requests time to close histogram")
+    time_close_histogram = px.histogram(df, x="timeToClose", title="Distribution of Time to Close Request", nbins=numBins, range_x=[min(
+        df.loc[:, "timeToClose"]), max(df.loc[:, "timeToClose"])], labels={"timeToClose": "Request Duration", "count": "Frequency"})
+    time_close_histogram.update_layout(margin=dict(l=50, r=50, b=50, t=50), font=dict(size=9))
+
+    # Time Series for the Total Number of Requests
+    print(" * Generating number of requests line chart")
+    req_time = pd.DataFrame(df.groupby("createDateDT", as_index=False)["srnumber"].count())
+    num_req_line_chart = px.line(req_time, x="createDateDT", y="srnumber", title="Total Number of 311 Requests Overtime", labels={
+                              "createDateDT": "DateTime", "srnumber": "Frequency"})
+    num_req_line_chart.update_layout(margin=dict(l=25, r=25, b=25, t=50), font=dict(size=9))
+
+    return req_type_pie_chart, time_close_histogram, num_req_line_chart, data_quality_output
+
+@callback(
+    Output("req_source_bar_chart", "figure"),
+    Output("req_source_bar_chart2", "figure"),
+    Output("total_req_card", "children"),
+    Output("total_req_card2", "children"),
+    Output("num_days_card", "children"),
+    Output("num_days_card2", "children"),
+    Output("overlay_req_time_line_chart", "figure"),
+    Input("nc_comp_dropdown", "value"),
+    Input("nc_comp_dropdown2", "value"),
+    prevent_initial_call=True
+)
+def generate_nc_comparison_charts(nc_comp_dropdown, nc_comp_dropdown2):
```

Comment:
Ditto--one callback per graph or visualization please.

Reply:
thanks for the suggestion - resolve.